File: coding-challenges/sources/leetcode/716. Max Stack/716-MaxStack.swift


fileprivate class MaxStack {
    private var mainStack: [Int]
    private var maxStack: [Int]

    init() {
        mainStack = []
        maxStack = []
    }

    func push(_ x: Int) {
        mainStack.append(x) // Push x onto mainStack
        // If maxStack is empty or x >= top of maxStack, push x onto maxStack
        if let lastMax = maxStack.last {
            if x >= lastMax {
                maxStack.append(x)
            } else {
                maxStack.append(lastMax)
            }
        } else {
            maxStack.append(x)
        }
    }

    func pop() -> Int {
        // If top of mainStack == top of maxStack, pop from maxStack
        if let lastMain = mainStack.last, let lastMax = maxStack.last, lastMain == lastMax {
            maxStack.removeLast()
        }
        // Pop and return the top element from mainStack
        return mainStack.removeLast()
    }

    func top() -> Int {
        return mainStack.last! // Return the top element of mainStack
    }

    func peekMax() -> Int {
        return maxStack.last! // Return the top element of maxStack
    }

    func popMax() -> Int {
        guard !maxStack.isEmpty else { return Int() }

        let maxValue = maxStack.removeLast() // Store the top element of maxStack as maxValue
        var tempStack: [Int] = [] // Create a temporary stack
        // While top of mainStack != maxValue, pop from mainStack and push the popped elements onto tempStack
        while mainStack.last! != maxValue {
            tempStack.append(pop())
        }
        _ = pop() // Pop from mainStack and maxStack
        // While tempStack is not empty, pop from tempStack and push the popped elements back onto mainStack,
        // following the push operation logic
        while !tempStack.isEmpty {
            push(tempStack.removeLast())
        }
        return maxValue
    }
}

import XCTest

final class MaxStackTests: XCTestCase {
    func test() {
        let maxStack = MaxStack()
        maxStack.push(5)
        maxStack.push(1)
        maxStack.push(5)
        XCTAssertEqual(maxStack.top(), 5)
        XCTAssertEqual(maxStack.popMax(), 5)
        XCTAssertEqual(maxStack.top(), 1)
        XCTAssertEqual(maxStack.peekMax(), 5)
        XCTAssertEqual(maxStack.pop(), 1)
        XCTAssertEqual(maxStack.top(), 5)
    }
}

File: coding-challenges/sources/leetcode/716. Max Stack/716-MaxStack.py
class MaxStack:

    def __init__(self):
        self.main_stack = []
        self.max_stack = []

    def push(self, x: int) -> None:
        self.main_stack.append(x) # push to main stack
        # push to max stack if max stack is empty or x is greater than or equal to the top of the max stack
        if not self.max_stack or x >= self.max_stack[-1]:
            self.max_stack.append(x)

    def pop(self) -> int:
        # if the top of the main stack is equal to the top of the max stack, pop from the max stack
        if self.main_stack[-1] == self.max_stack[-1]:
            self.max_stack.pop()
        # pop from the main stack
        return self.main_stack.pop()
    
    def top(self) -> int:
        return self.main_stack[-1] # return the top of the main stack
    
    def peekMax(self) -> int:
        return self.max_stack[-1] # return the top of the max stack
    
    def popMax(self) -> int:
        max_value = self.max_stack.pop() # pop from the max stack
        temp_stack = [] # create a temporary stack
        # while the top of the main stack is not equal to the max value, pop from the main stack and push to the temporary stack
        while self.main_stack[-1] != max_value:
            temp_stack.append(self.main_stack.pop())
        self.main_stack.pop() # pop from the main stack
        # while the temporary stack is not empty, pop from the temporary stack and push to the main stack
        while temp_stack:
            self.main_stack.append(temp_stack.pop())
        return max_value # return the max value

import unittest

class MaxStackTests(unittest.TestCase):
    def test(self):
        max_stack = MaxStack()
        max_stack.push(5)
        max_stack.push(1)
        max_stack.push(5)
        self.assertEqual(max_stack.top(), 5)
        self.assertEqual(max_stack.popMax(), 5)
        self.assertEqual(max_stack.top(), 1)
        self.assertEqual(max_stack.peekMax(), 5)
        self.assertEqual(max_stack.pop(), 1)
        self.assertEqual(max_stack.top(), 5)

if __name__ == '__main__':
    unittest.main()